tests/UserAgentsTest/FullTest.php

```diff
@@ -0,0 +1,240 @@
+    {
+        // First, generate the INI files
+        $buildNumber    = time();
+        $resourceFolder = __DIR__ . '/../../resources/';
+
+        self::$buildFolder = __DIR__ . '/../../build/browscap-ua-test-' . $buildNumber . '/build/';
+        $cacheFolder       = __DIR__ . '/../../build/browscap-ua-test-' . $buildNumber . '/cache/';
+
+        // create build folder if it does not exist
+        if (!file_exists(self::$buildFolder)) {
+            mkdir(self::$buildFolder, 0777, true);
+        }
+        if (!file_exists($cacheFolder)) {
+            mkdir($cacheFolder, 0777, true);
+        }
+
+        $logger = new Logger('browscap');
+        $logger->pushHandler(new NullHandler(Logger::DEBUG));
+
+        $buildGenerator = new BuildGenerator(
+            $resourceFolder,
+            self::$buildFolder
+        );
+
+        $buildGenerator->setLogger($logger);
+
+        $writerCollection = new WriterCollection();
+
+        self::$propertyHolder = new PropertyHolder();
+        self::$filter         = new FullFilter(self::$propertyHolder);
+
+        $fullPhpWriter = new IniWriter(self::$buildFolder . '/full_php_browscap.ini');
+        $formatter     = new PhpFormatter();
+        $formatter->setFilter(self::$filter);
+        $fullPhpWriter
+            ->setLogger($logger)
+            ->setFormatter($formatter)
+            ->setFilter(self::$filter);
+        $writerCollection->addWriter($fullPhpWriter);
+
+        $collectionCreator = new CollectionCreator();
+        $collectionCreator->setLogger($logger);
+
+        $buildGenerator->setCollectionCreator($collectionCreator);
+        $buildGenerator->setWriterCollection($writerCollection);
+        $buildGenerator->setCollectPatternIds(true);
+
+        $buildGenerator->run((string) $buildNumber, false);
+
+        $cache = new File([File::DIR => $cacheFolder]);
+        $cache->flush();
+
+        $resultFormatter = new LegacyFormatter();
+
+        self::$browscap = new Browscap();
+        self::$browscap
+            ->setCache($cache)
+            ->setLogger($logger)
+            ->setFormatter($resultFormatter);
+
+        self::$browscapUpdater = new BrowscapUpdater();
+        self::$browscapUpdater
+            ->setCache($cache)
+            ->setLogger($logger)
+            ->convertFile(self::$buildFolder . '/full_php_browscap.ini');
+    }
+
+    /**
+     * Runs after the entire test suite is run.  Generates a coverage report for JSON resource files if
+     * the $coveredPatterns array isn't empty
+     */
+    public static function tearDownAfterClass()
+    {
+        if (!empty(self::$coveredPatterns)) {
+            $coverageProcessor = new Processor(__DIR__ . '/../../resources/user-agents/');
+            $coverageProcessor->process(self::$coveredPatterns);
+            $coverageProcessor->write(__DIR__ . '/../../coverage-full.json');
+        }
+    }
+
+    /**
+     * @return array[]
+     */
+    public function userAgentDataProvider()
```

Comment:
Should this have the return type hint ` : array` or is that part of #1579?

Reply:
I will add all the type hint in PR #1579.